### PR TITLE
Rewrite line renderer: Part 1 (improved lasers plus other cleanups)

### LINKED ISSUE #7729: Laser beam start/end point "facing" extremely coarse-grained
Basically, the effect seems to know only four 'facings' (up, down, left, right), at least for the start and end of the line. Becomes very noticable with wider beams, see TS sonic tank for example.

<bountysource-plugin>
---
Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/9953771-laser-beam-start-end-point-facing-extremely-coarse-grained?utm_campaign=plugin&utm_content=tracker%2F36085&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F36085&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by pchote**
This introduces a new line renderer and uses it for:
- Widgets
- A\* / warhead debug overlays
- Range / detection circles
- Laser projectiles

Future PRs will work to remove the old line renderer completely, but this is already large enough and provides several benefits:
- Fixes #7729
- Fixes rendering of wide lines (like the TS disruptor) on some hardware (see discussion in #9357)
- We don't need to flush the renderer when changing line widths, so rendering range circles should in theory be more efficient
- Removes legacy QuadRenderer
- Implements a "flat beam" for use by the disruptor and sonic tank:

screen shot 2015-11-29 at 13 44 58

### CODE REVIEW

File: OpenRA.Game/Graphics/RgbaColorRenderer.cs [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+
+			vertices[nv++] = new Vertex(a + Offset, cr, cg, cb, ca);
+			vertices[nv++] = new Vertex(b + Offset, cr, cg, cb, ca);
+			vertices[nv++] = new Vertex(c + Offset, cr, cg, cb, ca);
+			vertices[nv++] = new Vertex(d + Offset, cr, cg, cb, ca);
+		}
+
+		public void FillEllipse(RectangleF r, Color color, int vertices = 32)
+		{
+			// TODO: Create an ellipse polygon instead
```
Comment (pchote) [inline OpenRA.Game/Graphics/RgbaColorRenderer.cs]: Resolving this depends on some changes that won't happen until part 2 or 3, so it just copies the (crap!) implementation from LineRenderer for now.

**Comment by pchote** [comment]
I'll deal with the style issue later.  That shouldn't impact the ability to test in the meantime.

File: OpenRA.Mods.Common/Traits/Render/RenderDetectionCircle.cs [hunk 2]
```diff
@@ -19,14 +19,11 @@ namespace OpenRA.Mods.Common.Traits
...
-		[Desc("Draw a rotating radar scanner update line, disabled by default.")]
-		public readonly bool DrawUpdateLine = false;
-
 		[Desc("WAngle the Radar update line advances per tick.")]
 		public readonly WAngle UpdateLineTick = new WAngle(-1);
 
-		[Desc("Number of trailing Radar update lines, will only draw one line if zero.")]
-		public readonly int LineTrailLength = 3;
+		[Desc("Number of trailing Radar update lines.")]
+		public readonly int TrailCount = 0;
```
Comment (abcdefg30) [inline OpenRA.Mods.Common/Traits/Render/RenderDetectionCircle.cs]: Won't this need an upgrade rule?

File: OpenRA.Mods.Common/Traits/Render/RenderDetectionCircle.cs [hunk 3]
```diff
@@ -19,14 +19,11 @@ namespace OpenRA.Mods.Common.Traits
 {
 	class RenderDetectionCircleInfo : ITraitInfo, Requires<DetectCloakedInfo>
 	{
-		[Desc("Draw a rotating radar scanner update line, disabled by default.")]
-		public readonly bool DrawUpdateLine = false;
```
Comment (abcdefg30) [inline OpenRA.Mods.Common/Traits/Render/RenderDetectionCircle.cs]: And this too? (Just to remove the junk values.)

**Comment by abcdefg30** [comment]
Derp. Nevermind, I guess those never went into a release version, so we don't need upgrade rules.

**Comment by pchote** [comment]
Fixed and rebased.

**Comment by UberWaffe** [comment]
Did a quick test with this rebased ontop AreaBeam in develop. Looks much better than the old rendering.

Will hopefully get time over the next few days to do a proper review.

**Comment by pchote** [comment]
Rebased.  This now includes the same rendering improvements for AreaBeam.

File: OpenRA.Game/Graphics/RgbaColorRenderer.cs [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+	public class RgbaColorRenderer : Renderer.IBatchRenderer
+	{
+		static readonly float2 Offset = new float2(0.5f, 0.5f);
+
+		readonly Renderer renderer;
+		readonly IShader shader;
+		readonly Action renderAction;
+
+		readonly Vertex[] vertices;
+		int nv = 0;
```
Comment (UberWaffe) [inline OpenRA.Game/Graphics/RgbaColorRenderer.cs]: Excuse my ignorance. What does `nv` stand for?

**Comment by UberWaffe** [comment]
Tentative :+1:
I did a test in TS, beam rendering for units (Obelisk, laser turret, disruptors, ghost and Mammoth Mk2) all seem fine.
Range circle rendering for cloak generator, obelisk, laser turret and sensor array all seem fine, and testing with cloak generator indicates the range correctly.
(I.e. I couldn't spot any ingame regressions).

I also don't see any problems with the code, but I'm also not familiar with the rendering code and what impacts this may have.

File: OpenRA.Game/Graphics/RgbaColorRenderer.cs (hunk 4 above)
Comment (pchote) [inline OpenRA.Game/Graphics/RgbaColorRenderer.cs]: number of vertices.  We use this abbreviation in a bunch of places in the low level rendering code.

Comment (Phrohdoh) [inline OpenRA.Game/Graphics/RgbaColorRenderer.cs]: I see no reason to be this terse if this has to be explained.
But changing that should be a separate PR.

File: OpenRA.Game/Graphics/RgbaColorRenderer.cs [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+			renderer.CurrentBatchRenderer = this;
+
+			if (nv + 4 > renderer.TempBufferSize)
+				Flush();
+
+			var delta = (end - start) / (end - start).Length;
+			var corner = width / 2 * new float2(-delta.Y, delta.X);
+
+			color = Util.PremultiplyAlpha(color);
+			var r = color.R / 255.0f;
```
Comment (Phrohdoh) [inline OpenRA.Game/Graphics/RgbaColorRenderer.cs]: It seems you are normalizing here, but what if the values have already been normalized?
"Don't do that" isn't valid IMO.

File: OpenRA.Game/Graphics/RgbaColorRenderer.cs [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+			vertices[nv++] = new Vertex(d + Offset, cr, cg, cb, ca);
+		}
+
+		public void FillEllipse(RectangleF r, Color color, int vertices = 32)
+		{
+			// TODO: Create an ellipse polygon instead
+			var a = (r.Right - r.Left) / 2;
+			var b = (r.Bottom - r.Top) / 2;
+			var xc = (r.Right + r.Left) / 2;
+			var yc = (r.Bottom + r.Top) / 2;
```
Comment (Phrohdoh) [inline OpenRA.Game/Graphics/RgbaColorRenderer.cs]: I assume `xc` and `yc` are x-change and y-change, in which case they should be `*-delta`: `xc, yc`.

File: OpenRA.Game/Graphics/RgbaColorRenderer.cs [hunk 7]
```diff
@@ -0,0 +1,129 @@
...
+			{
+				var dx = a * (float)Math.Sqrt(1 - (y - yc) * (y - yc) / b / b);
+				DrawLine(new float2(xc - dx, y), new float2(xc + dx, y), 1, color);
+			}
+		}
+
+		public void SetViewportParams(Size screen, float zoom, int2 scroll)
+		{
+			shader.SetVec("Scroll", scroll.X, scroll.Y);
+			shader.SetVec("r1", zoom * 2f / screen.Width, -zoom * 2f / screen.Height);
```
Comment (Phrohdoh) [inline OpenRA.Game/Graphics/RgbaColorRenderer.cs]: Can `r1` and `r2` be a bit more verbose?

File: OpenRA.Game/Graphics/RgbaColorRenderer.cs (hunk 5 above)
Comment (pchote) [inline OpenRA.Game/Graphics/RgbaColorRenderer.cs]: `System.Color` is not normalized.

File: OpenRA.Game/Graphics/RgbaColorRenderer.cs (hunk 6 above)
Comment (pchote) [inline OpenRA.Game/Graphics/RgbaColorRenderer.cs]: See comments (code + gh) above.  This implementation (which is moved from the old code) is going to be thrown out in the pr-after-next.

File: OpenRA.Game/Graphics/RgbaColorRenderer.cs (hunk 7 above)
Comment (pchote) [inline OpenRA.Game/Graphics/RgbaColorRenderer.cs]: Not without changing all of our shaders and renderers.

File: OpenRA.Mods.Common/Graphics/RangeCircleRenderable.cs [hunk 8]
```diff
@@ -43,13 +47,25 @@ public RangeCircleRenderable(WPos centerPosition, WDist radius, int zOffset, Col
...
-			wlr.LineWidth = 3;
-			wr.DrawRangeCircle(centerPosition, radius, contrastColor);
-			wlr.LineWidth = 1;
-			wr.DrawRangeCircle(centerPosition, radius, color);
-			wlr.LineWidth = oldWidth;
+			DrawRangeCircle(wr, centerPosition, radius, 1, color, 3, contrastColor);
+		}
+
+		public static void DrawRangeCircle(WorldRenderer wr, WPos centerPosition, WDist radius,
+			float width, Color color, float contrastWidth, Color contrastColor)
```
Comment (Phrohdoh) [inline OpenRA.Mods.Common/Graphics/RangeCircleRenderable.cs]: What is this the `width` of?
I imagine it is for the segmented lines.
Please be more verbose.

Comment (pchote) [inline OpenRA.Mods.Common/Graphics/RangeCircleRenderable.cs]: It is the width of the range circle that is being drawn by `DrawRangeCircle`.  I don't see how that can be more verbose without being redundant.

File: OpenRA.Mods.Common/Effects/AreaBeam.cs [hunk 9]
```diff
@@ -35,6 +35,9 @@ public class AreaBeamInfo : IProjectileInfo
 		[Desc("The width of the beam.")]
 		public readonly WDist Width = new WDist(512);
 
+		[Desc("The shape of the beam.  Accepts values Cylindrical or Flat.")]
```
Comment (Phrohdoh) [inline OpenRA.Mods.Common/Effects/AreaBeam.cs]: Maybe change Flat to Rectangular?

File: OpenRA.Game/Graphics/RgbaColorRenderer.cs (hunk 5 above)
Comment (Phrohdoh) [inline OpenRA.Game/Graphics/RgbaColorRenderer.cs]: This math could have been done manually and pushed back into the (new) Color, yes?

File: OpenRA.Mods.Common/Effects/AreaBeam.cs (hunk 9 above)
Comment (pchote) [inline OpenRA.Mods.Common/Effects/AreaBeam.cs]: It is flat on the ground, not rectangular:

File: OpenRA.Game/Graphics/RgbaColorRenderer.cs (hunk 5 above)
Comment (pchote) [inline OpenRA.Game/Graphics/RgbaColorRenderer.cs]: Not really... you would need to rewrite a bunch of our color/drawing code, and it would hurt performance.

**Comment by Mailaender** [comment]
Can't spot regressions on the touched parts. :white_check_mark: Only some transparency effects missing. Otherwise the TS beam weaponry seems to be finished. Nicely done!

**Comment by Mailaender** [comment]
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/c046638658dd53aa0b6a8a722a7e452a43ab0bfa...61898a7a8eb77990ab36988982bf832fccb7f21c)
